### DISCUSSION TITLE
Allow avatar_url to be set within the GitHub API

### DISCUSSION THREAD

**Discussion by anthr76** [community/community#13290] [Apps, API and Webhooks]
https://docs.github.com/en/rest/reference/teams#update-a-team

You cannot set an avatar_url within the update teams API. This request serves to request it. As orgs with zero-trust environments cannot easily update them.

**Comment by mehdichaouch** [comment]
@queenofcorgis I ping you to get help for this issue.
I have a pretty same needs, update the `avatar_url`, but of my account, and I saw nothing in [Users API](https://docs.github.com/en/rest/users/users#update-the-authenticated-user).
If you're not the right person could you please point out someone? Thanks

**Reply by queenofcorgis** [reply]
Hi there - looks like this was in the wrong category. The Issues category is solely for GitHub Issues and Projects products. I've moved it to the proper category so it will have better visibility among those who are able to assist.

**Reply by mehdichaouch** [reply]
@queenofcorgis thank you for helping.

**Reply by mehdichaouch** [reply]
@queenofcorgis :wave: - no update since 2 weeks, do you know someone who can help? Merci

**Comment by typed-sigterm** [comment]
bump

**Comment by AMDphreak** [comment]
This is still the right request, four years on. Please don't treat "use the website" as an answer.

**Ask:** `POST`/`PUT` to set avatars for users, orgs, teams, and GitHub Apps. Accept `multipart/form-data` or raw bytes on `uploads.github.com` (same path as release assets). Validate and transcode **on the server**. Optional crop box in the body. Return the new `avatar_url`.

### Why the usual objections don't hold

**1. "The REST API has no file upload / hosting has to stay in the browser."**
False. GitHub already uploads binaries over the API: release assets, repo file contents, Packages, Git LFS, gist files. The GHES 3.15 changelog even says secret scanning push protection is supported on **content upload REST API endpoints**. The org Settings form already POSTs a PNG into GitHub-hosted storage and `avatars.githubusercontent.com` serves a *re-encoded* blob (the bytes on the CDN are not the raw upload). An authenticated API is that same write with a token instead of a cookie.

**2. "Cropping, formatting, and moderation belong in the frontend."**
The cropper is UX, not a control. Anyone can bypass it with DevTools; the web UI is not a trust boundary. Your own docs already constrain PNG/JPG/GIF, <1 MB, <3000x3000, ~500x500 recommended. Enforce those with magic-byte checks, max pixels, and a server-side transcode/center-crop — you already re-encode. PhotoDNA / perceptual hashing / async classifiers belong **behind** that endpoint, not in JavaScript. Client-side moderation is theater: abuse already arrives through the same web form.

**3. "Accepting a customer-hosted avatar_url would hurt integrity."**
Then don't accept a hotlinked URL as the source of truth. Accept **bytes** and host them yourselves, exactly as the UI does. A one-shot fetch-and-copy (`logo_url` on App manifests, see #183826) is optional and is the same SSRF class you already handle for webhook URLs and Camo. Nobody who wants org-as-code is asking GitHub to permanently hotlink a random CDN.

**4. "Just use Gravatar."**
Gravatar is a third-party identity leak, doesn't cover teams or GitHub Apps, and the org "Gravatar email (Private)" field is not even on `PATCH /orgs/{org}`.

### Why this is no longer an edge case

- Org-as-code / [github-org-manager](https://github.com/OpenRailAssociation/github-org-manager) (#203949)
- Zero-trust shops that will not give a human a browser session against prod orgs (the original post here)
- GitHub Apps created via the manifest flow shipping with identicons (#183826)
- Agents and `gh` automation. GitHub is selling Copilot agents that only have the API. Leaving org identity as click-ops is a product hole in that world.

Gitea/Forgejo (`POST /orgs/{org}/avatar`), GitLab (multipart `avatar` on user/group update), and Bitbucket already do this. Identicons are generated server-side from an ID hash; replacing that blob is exposing a write the settings page already performs, not a new product surface.

Related: #65206 (users), #203949 (orgs/teams/apps; duplicate of this), #183826 (App logos).
